src/System.Linq.Parallel/tests/DegreeOfParallelismTests.cs

```diff
@@ -1,140 +1,86 @@
+                }
+            }
         }
 
-        [Fact]
+        [Theory]
+        [MemberData("DegreeData", new[] { 1024 }, new[] { 1, 4, 512 })]
         [OuterLoop]
-        public static void RunDopBarrierTest()
+        public static void DegreeOfParallelism(Labelled<ParallelQuery<int>> labelled, int count, int degree)
         {
-            RunDopBarrierTestCore(1);
-            RunDopBarrierTestCore(4);
-            RunDopBarrierTestCore(32);
-            // These variations take way too long and hit the 100s timeout.
-            //RunDopBarrierTest(64);
-            //RunDopBarrierTest(512);
+            Assert.Equal(Functions.SumRange(0, count), labelled.Item.WithDegreeOfParallelism(degree).Sum());
         }
 
-        [Fact]
+        [Theory]
+        [MemberData("DegreeData", new[] { 1, 4, 32 }, new object[] { })]
         [OuterLoop]
-        public static void RunDopPipeliningTest()
+        public static void DegreeOfParallelism_Barrier(Labelled<ParallelQuery<int>> labelled, int count, int degree)
         {
-            RunDopPipeliningTestCore(1000000, 1);
-            RunDopPipeliningTestCore(1000000, 4);
-            RunDopPipeliningTestCore(1000000, 64);
-            RunDopPipeliningTestCore(1000000, 512);
+            //int minWorker, minAsyncIO;
+            //ThreadPool.GetMinThreads(out minWorker, out minAsyncIO);
+            //ThreadPool.SetMinThreads(degree, minAsyncIO);
+            Barrier barrier = new Barrier(degree);
+            Assert.Equal(Functions.SumRange(0, count), labelled.Item.WithDegreeOfParallelism(degree).Sum(x => { barrier.SignalAndWait(); return x; }));
         }
 
-        [Fact]
+        [Theory]
+        [MemberData("DegreeData", new[] { 128 * 1024 }, new[] { 1, 4, 64, 512 })]
         [OuterLoop]
-        public static void RunDopPipeliningTestSleep()
+        public static void DegreeOfParallelism_Pipelining(Labelled<ParallelQuery<int>> labelled, int count, int degree)
         {
-            RunDopPipeliningTestSleepCore(1, 1);
-            RunDopPipeliningTestSleepCore(4, 4);
-            RunDopPipeliningTestSleepCore(32, 4);
-            //Re-enable for perf testing
-            //RunDopPipeliningTestSleep(64, 64);            
-            //RunDopPipeliningTestSleep(512, 512);
+            Assert.True(labelled.Item.WithDegreeOfParallelism(degree).Select(x => -x).OrderBy(x => x).SequenceEqual(ParallelEnumerable.Range(1 - count, count).AsOrdered()));
         }
 
-        //
-        // A simple test to check whether PLINQ appears to work with a particular DOP
-        //
-        private static void RunDopTest(int dop, bool expectException)
+        [Theory]
+        [MemberData("DegreeData", new[] { 1, 4 }, new object[] { })]
+        [MemberData("DegreeData", new[] { 32 }, new[] { 4 })]
+        // With more items and threads, the total runtime can far exceed the total delay.
+        // This does not appear to be solely due to thrashing - thread creation appears overly expensive.
```

Comment:
PLINQ doesn't create new threads, it queues tasks to run on the thread pool.  But the pool is relatively conservative about injecting new threads, so if lots of tasks are queued and all of those tasks need to run concurrently in order to make forward progress, the thread pool may only slowly ramp up to the required number of threads.